### DISCUSSION TITLE
link shortcuts are separated by a new lines

### DISCUSSION THREAD

**Discussion by muravjev** [remarkjs/remark#1068] [Q&A] [answered]
## Prerequisites

1. We have a link shortcuts at the end of a file:

```md
[link1]: any
[link2]: any
```

2. After running remark:

```sh
remark --output -- .
```

3. Newlines are added between shortcuts:

```md
[link1]: any

[link2]: any

```

diff: https://github.com/muravjev/repro_remark_shortcuts/compare/e6a60c6...2cf76bd

## Question

Is it possible to prevent such behavior?

## Reproduction

repo: https://github.com/muravjev/repro_remark_shortcuts

```sh
git clone https://github.com/muravjev/repro_remark_shortcuts.git
cd repro_remark_shortcuts
pnpm install
pnpm repro
```

**Comment by ChristianMurphy** [comment, marked as answer]
`remark` is an [abstract syntax tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree) tool.
Meaning it aims to preserve structure/meaning of content, but may not use the exact same characters.
Put another way it acts as a formatter.

The examples you provide are an example of this:

```markdown
[link1]: any
[link2]: any

and

[link1]: any

[link2]: any
```

are not directly rendered at all.
and only need to continue to exist somewhere, anywhere in the document, for it to be rendered the same.

Whether to always format with or without a newline is configurable with https://github.com/remarkjs/remark/tree/main/packages/remark-stringify#optionstightdefinitions

---

Though taking a step back, this does sound a bit like an [XY problem](https://xyproblem.info/).
What (if any) problem are the newlines creating?

Is another tool having trouble parsing them?
If so, that would be something to raise with that project, as it diverges from the [commonmark standard](https://commonmark.org), which much of the markdown community aims to adhere to.
